juju/juju
# [JUJU-1129] Refactor resource client api to be consistent with other juju client apis

The first on several PRs to bring resources into line with how facades and api clients are commonly implemented.
This PR starts with the api client layer. A lot of the changes are file moves and test fixes to use gomock. Some http helpers defined in api layer but used in apiserver layer were moved across. The agent api client is moved out to the agent sub-package.

## QA steps

bootstrap a machine controller
deploy the dummy-resource charm
use juju resources to see the resource metadata
It will result in a hook error because the bar resource is not supplied
Attach the missing resource
`juju attach-resource dummy-resource bar=baz.txt`
The status message now shows the resource has been downloaded with `resource-get`
ssh into the charm and inspect the resource content
use juju resources to inspect the resource metadata

bootstrap a k8s controller
deploy the mysql-k8s charm
It will stay allocating because there's no oci image in the store
Attach the mysql-image
`juju attach-resource mysql-k8s mysql-image=mariadb`
The charm will complete deployment
use juju resources to inspect the resource metadata
